networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py

```diff
@@ -0,0 +1,255 @@
+        from the Neutron (APIC ML2) workflow.
+        """
+        if not self._apic_driver:
+            try:
+                self._apic_driver = (
+                    manager.NeutronManager.get_service_plugins()[
+                        'GROUP_POLICY'].policy_driver_manager.policy_drivers[
+                            'apic'].obj)
+                self._get_vrf_context = self._get_vrf_context_gbp
+                self._get_ext_net_name = self._get_ext_net_name_gbp
+            except Exception:
+                pass
+            if not self._apic_driver:
+                try:
+                    self._apic_driver = (
+                        self._core_plugin.mechanism_manager.mech_drivers[
+                            'cisco_apic_ml2'].obj)
+                    self._get_vrf_context = self._get_vrf_context_ml2
+                    self._get_ext_net_name = self._get_ext_net_name_neutron
+                except Exception:
+                    pass
+        return self._apic_driver
+
+    def extend_hosting_port_info(self, context, port_db, hosting_device,
+                                 hosting_info):
+        """Get the segmenetation ID and interface
+
+        This extends the hosting info attribute with the segmentation ID
+        and physical interface used on the external router to connect to
+        the ACI fabric. The segmentation ID should have been set already
+        by the call to allocate_hosting_port, but if it's not present, use
+        the value from the port resource.
+        """
+        if hosting_info.get('segmentation_id') is None:
+            LOG.debug('No segmentation ID in hosting_info -- assigning')
+            hosting_info['segmentation_id'] = (
+                port_db.hosting_info.segmentation_id)
+        is_external = (port_db.device_owner == DEVICE_OWNER_ROUTER_GW)
+        hosting_info['physical_interface'] = self._get_interface_info(
+            hosting_device['id'], port_db.network_id, is_external)
+        ext_dict = self._get_external_network_dict(context, port_db)
+        if not is_external:
+            hosting_info['host_pool_cidr'] = ext_dict['host_pool_cidr']
+            hosting_info['next_hop'] = ext_dict['next_hop']
+            hosting_info['gateway_ip'] = ext_dict['gateway']
+
+    def allocate_hosting_port(self, context, router_id, port_db, network_type,
+                              hosting_device_id):
+        """Get the VLAN and port for this hosting device
+
+        The VLAN used between the APIC and the external router is stored
+        by the APIC driver.  This calls into the APIC driver to first get
+        the ACI VRF information associated with this port, then uses that
+        to look up the VLAN to use for this port to the external router
+        (kept as part of the L3 Out policy in ACI).
+        """
+        # If this is a router interface, the VLAN comes from APIC.
+        # If it's the gateway, the VLAN comes from the segment ID
```

Comment:
if its GW, the external network in ML2 (or even GBP maybe) is of type opflex which doesn't really allocate any vlan.....